File: fast-power/fast-power.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstring>
#include <cmath>
#include <utility>
#include <bitset>

#define mod7 1000000007

using namespace std;

long long fast_power_with_modulo(long long base, long long power, long long modulo)
{
    if(power == 0)
    {
        return 1LL;
    }
    else if(power == 1)
    {
        return base;
    }
    else
    {
        long long result = fast_power_with_modulo(base, power>>1, modulo);
        if(power&1)
        {
            return ((result*result%modulo)*base%modulo);
        }
        else
        {
            return (result*result%modulo);
        }
    }
}

long long fast_power(long long base, long long power)
{
    if(power == 0)
    {
        return 1LL;
    }
    else if(power == 1)
    {
        return base;
    }
    else
    {
        long long result = fast_power(base, power>>1);
        if(power&1)
        {
            return result * result * base;
        }
        else
        {
            return result * result;
        }
    }
}

int main()
{
    long long base = 2LL;
    long long power = 100LL;

    printf("%lld\n", fast_power_with_modulo(base,power, mod7));
    printf("%lld\n", fast_power(2,31));

    return 0;
}

File: fast-power/fast-power.rb
def fast_power_with_modulo(base, power, modulo)
    return 1 if power == 0
    return base if power == 1

    result = fast_power_with_modulo(base, power/2, modulo)
    return (result * result % modulo) * base % modulo if power%2 == 1
    return result * result % modulo
end

def fast_power(base, power)
    return 1 if power == 0
    return base if power == 1

    result = fast_power(base, power/2)
    return result * result * base if power%2 == 1
    return result * result
end

puts fast_power_with_modulo(2, 100, 1000000007)
puts fast_power(2, 31)